### PR TITLE
Allow xml docs to still be created when 'emit metadata only' is on.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
EmitMetadataOnly is a flag that exists for IDE scenarios to allow it to emit dlls cheaply that give hte top-level surface area of a particular compilation.  This can then be used in cross-language scenarios to allow things like intellisense/QI/and-lots-more to work in a live fashion.  We call these cross-language metadata-only dlls: 'skeleton assemblies'.

The IDE would like to improve our story around skeleton assemblies by being able to cache them to our persistence store so that later launched sessions of VS can reuse them without recomputing them.  

This is generally not difficult for us to do except for one small snag.  Specifically, one thing we support *today* with skeletons is being able to expose xml doc comments through the skeleton.  This works by having us actually both emit the metadta-only-dll, but also creating a DocumentationProvider that wraps the *original language* compilation.  So, for example, if VB references C#, it will get a dll for the C# metadata, but it will get a doc-provider that forwards along to the C# compilation to answer questions.

If we want to persist the dll (and still have xml-doc-comments work across languages) then we need to persist the xml doc info as well.  This PR updates the  compiler to support that.  Previously, the compiler would completely skip xml doc generation if EmitMetadataOnly was set.  With this PR, the compiler will emit xml doc comments as well if the appropriate stream is provided.

Relates to ref assemblies: https://github.com/dotnet/roslyn/issues/18612

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 1]
```diff
@@ -2662,6 +2665,12 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
                             ReportUnusedImports(diagnostics, cancellationToken);
                         }
                     }
+                    else if (xmlDocumentationStream != null)
+                    {
+                        // If we're in metadata only, and the caller asks for xml docs, then still proceed and generate // those.
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:2670]:
```suggestion
                        // If we're in metadata only, and the caller asks for xml docs, then still proceed and generate those.
```

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  ptal :)

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 2]
```diff
@@ -3306,11 +3306,16 @@ private void GenerateModuleInitializer(PEModuleBuilder moduleBeingBuilt, Diagnos
...
 
-            if (!FilterAndAppendAndFreeDiagnostics(diagnostics, ref resourceDiagnostics, cancellationToken))
-            {
-                return false;
-            }
+            return FilterAndAppendAndFreeDiagnostics(diagnostics, ref resourceDiagnostics, cancellationToken);
+        }
 
+        internal override bool GenerateDocumentationComments(
+            CommonPEModuleBuilder moduleBuilder,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:3313]: Is this parameter used?

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 3]
```diff
@@ -246,6 +245,241 @@ public static void Main()
...
+            {
+                emitResult = comp.Emit(peStream, xmlDocumentationStream: xmlStream, options: new EmitOptions(metadataOnly: true));
+                mdOnlyImage = peStream.ToArray();
+                xmlDocBytes = xmlStream.ToArray();
+            }
+
+            Assert.True(emitResult.Success);
+            emitResult.Diagnostics.Verify();
+            Assert.True(mdOnlyImage.Length > 0, "no metadata emitted");
+            Assert.True(xmlDocBytes.Length > 0, "no xml emitted");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:280]: Consider verifying that "This should be emitted" made it into the output. `Length > 0` is not a strong test.

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 4]
```diff
@@ -246,6 +245,241 @@ public static void Main()
             }
         }
 
+        [Fact]
+        public void EmitMetadataOnly_XmlDocs_NoDocMode_Success()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:249]: What does "NoDocMode" mean?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 7)

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:249]: SureSure!

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:3313]: nope!  removing. thanks :)

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:249]: done.

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 5]
```diff
@@ -246,6 +246,253 @@ public static void Main()
...
+        {
+            CSharpCompilation comp = CreateCompilation(@"
+namespace Goo.Bar
+{
+    /// <summary>This should be emitted</summary>
+    public class Test1
+    {
+        public static void SayHello()
+        {
+            ConEmitMetadataOnly_XmlDocs_NoDocMode_Successsole.WriteLine(""hello"");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:260]: >ConEmitMetadataOnly_XmlDocs_NoDocMode_Successsole [](http://example.com/codeflow?start=12&length=49)

Accidental paste?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 10) modulo typo in test

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 6]
```diff
@@ -246,6 +246,253 @@ public static void Main()
...
+
+            using (var peStream = new MemoryStream())
+            using (var xmlStream = new MemoryStream())
+            {
+                emitResult = comp.Emit(peStream, xmlDocumentationStream: xmlStream, options: new EmitOptions(metadataOnly: true));
+                mdOnlyImage = peStream.ToArray();
+                xmlDocBytes = xmlStream.ToArray();
+            }
+
+            // This should not fail the emit (as it's a warning).
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:485]: nit: Please remove this comment here and the one below `// Even though docs failed, ...`

**Review by RikkiGibson — APPROVED** [review]
LGTM except the feedback above (just have to address before merge).

**Comment by jcouv** [comment]
Feel free to merge/squash once the typo and last comment are addressed/considered. Thanks

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:260]:
```suggestion
            EmitMetadataOnly_XmlDocs_NoDocMode_Success.WriteLine(""hello"");
```

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:485]:
```suggestion
```

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 7]
```diff
@@ -246,6 +246,252 @@ public static void Main()
...
+            {
+                emitResult = comp.Emit(peStream, xmlDocumentationStream: xmlStream, options: new EmitOptions(metadataOnly: true));
+                mdOnlyImage = peStream.ToArray();
+                xmlDocBytes = xmlStream.ToArray();
+            }
+
+            Assert.True(emitResult.Success);
+            emitResult.Diagnostics.Verify();
+
+            // Even though docs failed, we should still produce the peStream.
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:488]:
```suggestion
```

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: This doesn't look right, aren't we trying to call Console.WriteLine here?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: sigh...

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 8]
```diff
@@ -246,6 +246,251 @@ public static void Main()
...
+        {
+            CSharpCompilation comp = CreateCompilation(@"
+namespace Goo.Bar
+{
+    /// <summary>This should be emitted</summary>
+    public class Test1
+    {
+        public static void SayHello()
+        {
+            EmitMetadataOnly_XmlDocs_NoDocMode_Success.WriteLine(""hello"");
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:260]:
```suggestion
            Console.WriteLine(""hello"");
```

**Pull request merged**